h11/_receivebuffer.py

```diff
@@ -38,75 +40,101 @@
-        # Returns extracted bytes on success (advancing offset), or None on
-        # failure
-        if self._looked_for == needle:
-            search_start = max(self._start, self._looked_at - len(needle) + 1)
-        else:
-            search_start = self._start
-        offset = self._data.find(needle, search_start)
-        if offset == -1:
-            self._looked_at = len(self._data)
-            self._looked_for = needle
+    def maybe_extract_next_line(self):
+        """
+        Extract the first line, if it is completed in the buffer.
+        """
+        # Only search in buffer space that we've not already looked at.
+        search_start_index = max(0, self._next_line_search - 1)
+        partial_buffer = self._data[search_start_index:]
+        partial_idx = partial_buffer.find(b"\r\n")
+        if partial_idx == -1:
+            self._next_line_search = len(self._data)
             return None
-        new_start = offset + len(needle)
-        out = self._data[self._start : new_start]
-        self._start = new_start
+
+        # Truncate the buffer and return it.
+        # + 2 is to compensate len(b"\r\n")
+        idx = search_start_index + partial_idx + 2
+        out = self._data[:idx]
+        self._data[:idx] = b""
+        self._next_line_search = 0
+        self._multiple_lines_search = 0
         return out
 
-    # HTTP/1.1 has a number of constructs where you keep reading lines until
-    # you see a blank one. This does that, and then returns the lines.
     def maybe_extract_lines(self):
-        if self._data[self._start : self._start + 2] == b"\r\n":
-            self._start += 2
+        """
+        Extract everything up to the first blank line, and return a list of lines.
+        """
+        # Handle the case where we have an immediate empty line.
+        if self._data[:1] == b"\n":
+            self._data[:1] = b""
+            self._next_line_search = 0
+            self._multiple_lines_search = 0
+            return []
+
+        if self._data[:2] == b"\r\n":
+            self._data[:2] = b""
+            self._next_line_search = 0
+            self._multiple_lines_search = 0
             return []
-        else:
-            data = self.maybe_extract_until_next(b"\r\n\r\n")
-            if data is None:
-                return None
-            lines = data.split(b"\r\n")
-            assert lines[-2] == lines[-1] == b""
-            del lines[-2:]
-            return lines
+
+        # Only search in buffer space that we've not already looked at.
+        partial_buffer = self._data[self._multiple_lines_search :]
+        match = blank_line_regex.search(partial_buffer)
+        if match is None:
+            self._multiple_lines_search = max(0, len(self._data) - 2)
+            return None
+
+        # Truncate the buffer and return it.
+        idx = self._multiple_lines_search + match.span(0)[-1]
+        out = self._data[:idx]
+        lines = [line.rstrip(b"\r") for line in out.split(b"\n")]
```

Comment:
Instead of calling `rstrip` here (which always has to copy the whole buffer, since these are bytearrays), I think we could leave the trailing `\r` in, and then in `_abnf.py` change `header_field`, `request_line`, and `status_line` to match a trailing optional `\r`, e.g.:

```python
header_field = (
    r"(?P<field_name>{field_name})"
    r":"
    r"{OWS}"
    r"(?P<field_value>{field_value})"
    r"{OWS}\r?".format(**globals())         # <-- notice added \r? here
)
```

Reply:
Hello @njsmith ,

I tried to do that.

I added `\r?` to these regexes (`header_field`, and `request_line`, and `status_line`)

Then I set
```python
lines = out.split(b"\n")
```

but it broke one of `test_readers_unusual` test cases in `test_io.py`
<details>
  <summary>the test source</summary>

```python
def test_readers_unusual():
    ...

    # obsolete line folding
    tr(
        READERS[CLIENT, IDLE],
        b"HEAD /foo HTTP/1.1\r\n"
        b"Host: example.com\r\n"
        b"Some: multi-line\r\n"
        b" header\r\n"
        b"\tnonsense\r\n"
        b"    \t   \t\tI guess\r\n"
        b"Connection: close\r\n"
        b"More-nonsense: in the\r\n"
        b"    last header  \r\n\r\n",
        Request(
            method="HEAD",
            target="/foo",
            headers=[
                ("Host", "example.com"),
                ("Some", "multi-line header nonsense I guess"),
                ("Connection", "close"),
                ("More-nonsense", "in the last header"),
            ],
        ),
    )
``` 

</details>

The header
```python
        b"Some: multi-line\r\n"
        b" header\r\n"
        b"\tnonsense\r\n"
        b"    \t   \t\tI guess\r\n"
```
turns to
```python
        b"Some: multi-line\r header\r\tnonsense\r    \t   \t\tI guess\r\n"
```

I cannot figure out how to carefully cut out the `\r` from such a line.

Reply:
After some discussion in [gitter](https://gitter.im/python-hyper/community?at=5fd7843082222960e4e589fd), it has been decided not to change regexes, but to rewrite `.rstrip(...)` with 
```python
        for line in lines:
            if line.endswith(b"\r"):
                del line[-1]
```
to avoid extra memory allocation